### ISSUE TITLE
Cannot install packages on Mac

### ISSUE THREAD

**Issue by sminneman** [GlitchEnzo/NuGetForUnity#374]
## Description
When trying to install packages on my Mac, it keeps coming up with System.UnauthorizedAccessException: Access to the path [snipped] is denied.

I have tried multiple versions of the Editor, multiple locations of the Editor, and multiple packages.

I have verified that I can access the path listed, and can create files/folders in that path.

I have tried multiple NuGet packages.

Interestingly, I can Update packages that came with NuGetForUnity just fine. I just cannot install new ones.

I tried searching for other examples of people having this issue, but could not find any - apologies if this has been asked/answered and I overlooked it.

Any ideas? Thanks everyone!

- **NuGet Package:** Microsoft.Bot.Builder 4.11.1
- **NuGetForUnity Version:** 3.0.1
- **Unity Version:** 2020.1.17f1
- **Operating System:** MacOS Big Sur, 11.2

**Comment by KKKKKyle** [comment]
Experiencing the same issue here. I was trying to install the package "Google.Cloud.PubSub.V1" and I got stuck at one of its dependencies "Grpc.Core 2.36.0". Here's the full error output:

> Unable to install package Grpc.Core 2.36.0
> System.UnauthorizedAccessException: Access to the path 'MyProjectPath/Assets/Packages/Grpc.Core.2.36.0/build/' is denied.
>   at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean anonymous, System.IO.FileOptions options) [0x000e0] in <9577ac7a62ef43179789031239ba8798>:0 
>   at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.IO.FileOptions options) [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
>   at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean useAsync) [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
>   at (wrapper remoting-invoke-with-check) System.IO.FileStream..ctor(string,System.IO.FileMode,System.IO.FileAccess,System.IO.FileShare,int,bool)
>   at System.IO.Compression.ZipFileExtensions.ExtractToFile (System.IO.Compression.ZipArchiveEntry source, System.String destinationFileName, System.Boolean overwrite) [0x00024] in <21010871e8cb4079b5e9593dc3e69066>:0 
>   at NugetForUnity.NugetHelper.Install (NugetForUnity.NugetPackage package, System.Boolean refreshAssets) [0x003fe] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
> UnityEngine.Debug:LogErrorFormat(String, Object[])
> NugetForUnity.NugetHelper:Install(NugetPackage, Boolean)
> NugetForUnity.NugetHelper:InstallIdentifier(NugetPackageIdentifier, Boolean)
> NugetForUnity.NugetHelper:Restore()
> NugetForUnity.NugetWindow:RestorePackages()

NuGetForUnity Version: 3.0.1
Unity Version: 2019.4.15f1
OS: MacOS Catalina

Thanks in advance!

**Comment by CodeMasterYi** [comment]
Also on Mac...
```
Unable to retrieve package list from http://www.nuget.org/api/v2/Search()?$filter=IsLatestVersion&$orderby=DownloadCount desc&$skip=0&$top=15&searchTerm=''&targetFramework=''&includePrerelease=false
System.Xml.XmlException: Unexpected end of file while parsing Name has occurred. Line 2, position 1703.
  at System.Xml.XmlTextReaderImpl.Throw (System.Exception e) [0x00027] in <7fd195060d8c41448694ab221d3b56ca>:0 
  at System.Xml.XmlTextReaderImpl.Throw (System.String res, System.String arg) [0x00029] in <7fd195060d8c41448694ab221d3b56ca>:0 
  at System.Xml.XmlTextReaderImpl.Throw (System.Int32 pos, System.String res, System.String arg) [0x0000c] in <7fd195060d8c41448694ab221d3b56ca>:0 
  at System.Xml.XmlTextReaderImpl.ParseQName (System.Boolean isQName, System.Int32 startOffset, System.Int32& colonPos) [0x00124] in <7fd195060d8c41448694ab221d3b56ca>:0 
  at System.Xml.XmlTextReaderImpl.ParseQName (System.Int32& colonPos) [0x00000] in <7fd195060d8c41448694ab221d3b56ca>:0 
  at System.Xml.XmlTextReaderImpl.ThrowTagMismatch (System.Xml.XmlTextReaderImpl+NodeData startTag) [0x0000c] in <7fd195060d8c41448694ab221d3b56ca>:0 
  at System.Xml.XmlTextReaderImpl.ParseEndElement () [0x000d7] in <7fd195060d8c41448694ab221d3b56ca>:0 
  at System.Xml.XmlTextReaderImpl.ParseElementContent () [0x00149] in <7fd195060d8c41448694ab221d3b56ca>:0 
  at System.Xml.XmlTextReaderImpl.Read () [0x00085] in <7fd195060d8c41448694ab221d3b56ca>:0 
  at System.Xml.Linq.XContainer.ReadContentFrom (System.Xml.XmlReader r) [0x00201] in <428ab70c13974bd0bcd1eeafc00ea3e2>:0 
  at System.Xml.Linq.XContainer.ReadContentFrom (System.Xml.XmlReader r, System.Xml.Linq.LoadOptions o) [0x00005] in <428ab70c13974bd0bcd1eeafc00ea3e2>:0 
  at System.Xml.Linq.XDocument.Load (System.Xml.XmlReader reader, System.Xml.Linq.LoadOptions options) [0x00080] in <428ab70c13974bd0bcd1eeafc00ea3e2>:0 
  at System.Xml.Linq.XDocument.Load (System.IO.TextReader textReader, System.Xml.Linq.LoadOptions options) [0x0000f] in <428ab70c13974bd0bcd1eeafc00ea3e2>:0 
  at System.Xml.Linq.XDocument.Load (System.IO.TextReader textReader) [0x00000] in <428ab70c13974bd0bcd1eeafc00ea3e2>:0 
  at NugetForUnity.NugetPackageSource.GetPackagesFromUrl (System.String url, System.String username, System.String password) [0x00079] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
  at NugetForUnity.NugetPackageSource.Search (System.String searchTerm, System.Boolean includeAllVersions, System.Boolean includePrerelease, System.Int32 numberToGet, System.Int32 numberToSkip) [0x000ce] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
UnityEngine.Debug:LogErrorFormat(String, Object[])
NugetForUnity.NugetPackageSource:Search(String, Boolean, Boolean, Int32, Int32)
NugetForUnity.NugetHelper:Search(String, Boolean, Boolean, Int32, Int32)
NugetForUnity.NugetWindow:UpdateOnlinePackages()
NugetForUnity.NugetWindow:Refresh(Boolean)
NugetForUnity.NugetWindow:OnEnable()
UnityEditor.EditorWindow:GetWindow()
NugetForUnity.NugetWindow:DisplayNugetWindow()
```

**Comment by CodeMasterYi** [comment]
And this error does not exist on Windows.

**Comment by jwittner** [comment]
Unfortunately I don't have a Mac to test/fix these issues. If someone in the community wants to pick up debugging/fixing these on Mac I'd gladly review and test on Windows to ensure no regressions there.

**Comment by CodeMasterYi** [comment]
> Unfortunately I don't have a Mac to test/fix these issues. If someone in the community wants to pick up debugging/fixing these on Mac I'd gladly review and test on Windows to ensure no regressions there.

you can give me a new package to test it.

**Comment by KKKKKyle** [comment]
@jwittner I actually had the same issue when I tried to install Google.Cloud.Pubsub.V1 on Windows. I must manually change the access of downloaded packages by unchecking "Read Only" (though there's another issue that stops me from successfully installing it though, which may relate to Google's Protobuf package itself).

**Comment by cgleadall** [comment]
> And this error does not exist on Windows.

I have this same issue on Windows 10. I can't install Grpc.Net.Client or Google.Cloud.Storage.V1

I also see the same error on a MAC, as the op.

@CodeMasterYi Can you please let us know what versions you are on to allow you to install `Grpc.Net.Client` in your project on Windows?

This excerpt is from clicking 'Upgrade' on a previously installed package that has an update.
```
Unable to install package Grpc.Core.Api 2.36.4
System.UnauthorizedAccessException: Access to the path 'C:\Users\me\...\Packages\Grpc.Core.Api.2.36.4\lib\' is denied.
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean anonymous, System.IO.FileOptions options) [0x000e0] in <eae584ce26bc40229c1b1aa476bfa589>:0 
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.IO.FileOptions options) [0x00000] in <eae584ce26bc40229c1b1aa476bfa589>:0 
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean useAsync) [0x00000] in <eae584ce26bc40229c1b1aa476bfa589>:0 
  at (wrapper remoting-invoke-with-check) System.IO.FileStream..ctor(string,System.IO.FileMode,System.IO.FileAccess,System.IO.FileShare,int,bool)
  at System.IO.Compression.ZipFileExtensions.ExtractToFile (System.IO.Compression.ZipArchiveEntry source, System.String destinationFileName, System.Boolean overwrite) [0x00024] in <21010871e8cb4079b5e9593dc3e69066>:0 
  at NugetForUnity.NugetHelper.Install (NugetForUnity.NugetPackage package, System.Boolean refreshAssets) [0x003fe] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
UnityEngine.Debug:LogErrorFormat (string,object[])
NugetForUnity.NugetHelper:Install (NugetForUnity.NugetPackage,bool)
NugetForUnity.NugetHelper:InstallIdentifier (NugetForUnity.NugetPackageIdentifier,bool)
NugetForUnity.NugetHelper:Update (NugetForUnity.NugetPackageIdentifier,NugetForUnity.NugetPackage,bool)
NugetForUnity.NugetWindow:DrawPackage (NugetForUnity.NugetPackage,UnityEngine.GUIStyle,UnityEngine.GUIStyle)
NugetForUnity.NugetWindow:DrawPackages (System.Collections.Generic.List`1<NugetForUnity.NugetPackage>)
NugetForUnity.NugetWindow:DrawUpdates ()
NugetForUnity.NugetWindow:OnGUI ()
UnityEngine.GUIUtility:ProcessEvent (int,intptr)
```

Windows: 10
Version:  20H2
Build 19042.867

Unity: 2019.4.21f1

**Comment by BillG212** [comment]
Also received the below four errors  while trying to install CsvHelper on Mac.

````
Unable to retrieve package list from http://www.nuget.org/api/v2/FindPackagesById()?id='System.Threading.Tasks.Extensions'&$orderby=Version asc
System.Net.WebException: The remote server returned an error: (400) Bad Request.
  at System.Net.HttpWebRequest.EndGetResponse (System.IAsyncResult asyncResult) [0x00058] in <ef151b6abb5d474cb2c1cb8906a8b5a4>:0 
  at System.Net.HttpWebRequest.GetResponse () [0x0000e] in <ef151b6abb5d474cb2c1cb8906a8b5a4>:0 
  at NugetForUnity.NugetHelper.RequestUrl (System.String url, System.String userName, System.String password, System.Nullable`1[T] timeOut) [0x000b9] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
  at NugetForUnity.NugetPackageSource.GetPackagesFromUrl (System.String url, System.String username, System.String password) [0x0006c] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
  at NugetForUnity.NugetPackageSource.FindPackagesById (NugetForUnity.NugetPackageIdentifier package) [0x000ee] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
UnityEngine.Debug:LogErrorFormat (string,object[])
NugetForUnity.NugetPackageSource:FindPackagesById (NugetForUnity.NugetPackageIdentifier)
NugetForUnity.NugetPackageSource:GetSpecificPackage (NugetForUnity.NugetPackageIdentifier)
NugetForUnity.NugetHelper:GetOnlinePackage (NugetForUnity.NugetPackageIdentifier)
NugetForUnity.NugetHelper:GetSpecificPackage (NugetForUnity.NugetPackageIdentifier)
NugetForUnity.NugetHelper:InstallIdentifier (NugetForUnity.NugetPackageIdentifier,bool)
NugetForUnity.NugetHelper:Install (NugetForUnity.NugetPackage,bool)
NugetForUnity.NugetHelper:InstallIdentifier (NugetForUnity.NugetPackageIdentifier,bool)
NugetForUnity.NugetHelper:Install (NugetForUnity.NugetPackage,bool)
NugetForUnity.NugetHelper:InstallIdentifier (NugetForUnity.NugetPackageIdentifier,bool)
NugetForUnity.NugetWindow:DrawPackage (NugetForUnity.NugetPackage,UnityEngine.GUIStyle,UnityEngine.GUIStyle)
NugetForUnity.NugetWindow:DrawPackages (System.Collections.Generic.List`1<NugetForUnity.NugetPackage>)
NugetForUnity.NugetWindow:DrawOnline ()
NugetForUnity.NugetWindow:OnGUI ()
UnityEngine.GUIUtility:ProcessEvent (int,intptr) (at /Users/bokken/buildslave/unity/build/Modules/IMGUI/GUIUtility.cs:197)
```

`Could not find System.Threading.Tasks.Extensions [4.5.2, ) or greater.
UnityEngine.Debug:LogErrorFormat (string,object[])
NugetForUnity.NugetHelper:InstallIdentifier (NugetForUnity.NugetPackageIdentifier,bool)
NugetForUnity.NugetHelper:Install (NugetForUnity.NugetPackage,bool)
NugetForUnity.NugetHelper:InstallIdentifier (NugetForUnity.NugetPackageIdentifier,bool)
NugetForUnity.NugetHelper:Install (NugetForUnity.NugetPackage,bool)
NugetForUnity.NugetHelper:InstallIdentifier (NugetForUnity.NugetPackageIdentifier,bool)
NugetForUnity.NugetWindow:DrawPackage (NugetForUnity.NugetPackage,UnityEngine.GUIStyle,UnityEngine.GUIStyle)
NugetForUnity.NugetWindow:DrawPackages (System.Collections.Generic.List`1<NugetForUnity.NugetPackage>)
NugetForUnity.NugetWindow:DrawOnline ()
NugetForUnity.NugetWindow:OnGUI ()
UnityEngine.GUIUtility:ProcessEvent (int,intptr) (at /Users/bokken/buildslave/unity/build/Modules/IMGUI/GUIUtility.cs:197)`

`Unable to install package Microsoft.Bcl.AsyncInterfaces 1.1.0
System.Exception: Failed to install dependency: System.Threading.Tasks.Extensions [4.5.2, ).
  at NugetForUnity.NugetHelper.Install (NugetForUnity.NugetPackage package, System.Boolean refreshAssets) [0x001a1] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
UnityEngine.Debug:LogErrorFormat (string,object[])
NugetForUnity.NugetHelper:Install (NugetForUnity.NugetPackage,bool)
NugetForUnity.NugetHelper:InstallIdentifier (NugetForUnity.NugetPackageIdentifier,bool)
NugetForUnity.NugetHelper:Install (NugetForUnity.NugetPackage,bool)
NugetForUnity.NugetHelper:InstallIdentifier (NugetForUnity.NugetPackageIdentifier,bool)
NugetForUnity.NugetWindow:DrawPackage (NugetForUnity.NugetPackage,UnityEngine.GUIStyle,UnityEngine.GUIStyle)
NugetForUnity.NugetWindow:DrawPackages (System.Collections.Generic.List`1<NugetForUnity.NugetPackage>)
NugetForUnity.NugetWindow:DrawOnline ()
NugetForUnity.NugetWindow:OnGUI ()
UnityEngine.GUIUtility:ProcessEvent (int,intptr) (at /Users/bokken/buildslave/unity/build/Modules/IMGUI/GUIUtility.cs:197)`

`Unable to install package CsvHelper 26.1.0
System.Exception: Failed to install dependency: Microsoft.Bcl.AsyncInterfaces 1.1.0.
  at NugetForUnity.NugetHelper.Install (NugetForUnity.NugetPackage package, System.Boolean refreshAssets) [0x001a1] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
UnityEngine.Debug:LogErrorFormat (string,object[])
NugetForUnity.NugetHelper:Install (NugetForUnity.NugetPackage,bool)
NugetForUnity.NugetHelper:InstallIdentifier (NugetForUnity.NugetPackageIdentifier,bool)
NugetForUnity.NugetWindow:DrawPackage (NugetForUnity.NugetPackage,UnityEngine.GUIStyle,UnityEngine.GUIStyle)
NugetForUnity.NugetWindow:DrawPackages (System.Collections.Generic.List`1<NugetForUnity.NugetPackage>)
NugetForUnity.NugetWindow:DrawOnline ()
NugetForUnity.NugetWindow:OnGUI ()
UnityEngine.GUIUtility:ProcessEvent (int,intptr) (at /Users/bokken/buildslave/unity/build/Modules/IMGUI/GUIUtility.cs:197)`

**Comment by jwittner** [comment]
@BelalGawish - you're issue has to do with OData APIs being deprecated. This has been fixed in latest commits, but not published yet. Waiting for some help to review #385 so that I can publish both sets of fixes at once.

**Comment by BelalGawish** [comment]
@jwittner ok Thanks. I will wait for it.

**Comment by kyjanond** [comment]
Hi. I have same issue over here. Any progress on the release?

**Comment by jwittner** [comment]
Changes have been published as [Release 3.0.2](https://github.com/GlitchEnzo/NuGetForUnity/releases/tag/v3.0.2).
